File: src/collab/remove.py
"""
Remove a collaborator: deinit submodule, remove from config.

Usage:
  corrkit for remove alex-gh
  corrkit for remove alex-gh --delete-repo   # Also delete the GitHub repo
"""

import argparse
import subprocess
import sys
from pathlib import Path

from . import collab_dir, load_collaborators, save_collaborators


def _run(cmd: list[str], check: bool = True) -> subprocess.CompletedProcess:
    print(f"  $ {' '.join(cmd)}")
    result = subprocess.run(cmd, capture_output=True, text=True)
    if check and result.returncode != 0:
        print(f"  FAILED: {result.stderr.strip()}")
    return result


def main() -> None:
    parser = argparse.ArgumentParser(description="Remove a collaborator")
    parser.add_argument("name", help="Collaborator GitHub username to remove")
    parser.add_argument(
        "--delete-repo",
        action="store_true",
        help="Also delete the GitHub repo (requires confirmation)",
    )
    args = parser.parse_args()

    name: str = args.name
    collabs = load_collaborators()

    if name not in collabs:
        print(f"Collaborator '{name}' not found in collaborators.toml")
        sys.exit(1)

    collab = collabs[name]
    sub_path = collab_dir(collab)

    # 1. Deinit and remove submodule
    if sub_path.exists():
        print(f"Removing submodule: {sub_path}")
        _run(["git", "submodule", "deinit", "-f", str(sub_path)])
        _run(["git", "rm", "-f", str(sub_path)])
    else:
        print(f"Submodule {sub_path} not found on disk -- skipping git cleanup")

    # Clean up .git/modules entry
    modules_path = Path(".git/modules") / str(sub_path)
    if modules_path.exists():
        _run(["rm", "-rf", str(modules_path)])

    # 2. Remove from collaborators.toml
    del collabs[name]
    save_collaborators(collabs)
    print(f"Removed '{name}' from collaborators.toml")

    # 3. Optionally delete the GitHub repo
    if args.delete_repo:
        confirm = input(
            f"Delete GitHub repo {collab.repo}? This cannot be undone. [y/N] "
        )
        if confirm.lower() == "y":
            _run(["gh", "repo", "delete", collab.repo, "--yes"])
            print(f"Deleted GitHub repo: {collab.repo}")
        else:
            print("Skipped repo deletion")

    print(f"Done. Collaborator '{name}' removed.")


if __name__ == "__main__":
    main()

File: src/mailbox/remove.rs
//! Remove a mailbox: plain directory or submodule.

use anyhow::Result;
use std::io::Write;
use std::path::PathBuf;

use crate::resolve;
use crate::util::run_cmd_checked;

pub fn run(name: &str, delete_repo: bool) -> Result<()> {
    let mb_path = resolve::mailbox_dir(name);

    if mb_path.exists() {
        let is_submodule = mb_path.join(".git").is_file();

        if is_submodule {
            // Submodule removal
            println!("Removing submodule: {}", mb_path.display());
            let sp = mb_path.to_string_lossy().to_string();
            run_cmd_checked(&["git", "submodule", "deinit", "-f", &sp])?;
            run_cmd_checked(&["git", "rm", "-f", &sp])?;

            // Clean up .git/modules entry
            let modules_path =
                PathBuf::from(".git/modules").join(mb_path.to_string_lossy().as_ref());
            if modules_path.exists() {
                std::fs::remove_dir_all(&modules_path)?;
                println!("  Cleaned up {}", modules_path.display());
            }
        } else {
            // Plain directory removal
            println!("Removing directory: {}", mb_path.display());
            std::fs::remove_dir_all(&mb_path)?;
        }
    } else {
        println!(
            "Mailbox {} not found on disk -- skipping cleanup",
            mb_path.display()
        );
    }

    // Remove from .corky.toml
    remove_from_config(name)?;

    // Optionally delete GitHub repo
    if delete_repo {
        // Try to find repo name from config or convention
        let owner = crate::accounts::load_owner(None).ok();
        let repo_full = owner
            .map(|o| format!("{}/to-{}", o.github_user, name.to_lowercase()))
            .unwrap_or_default();

        if !repo_full.is_empty() {
            print!(
                "Delete GitHub repo {}? This cannot be undone. [y/N] ",
                repo_full
            );
            std::io::stdout().flush()?;
            let mut input = String::new();
            std::io::stdin().read_line(&mut input)?;
            if input.trim().to_lowercase() == "y" {
                run_cmd_checked(&["gh", "repo", "delete", &repo_full, "--yes"])?;
                println!("Deleted GitHub repo: {}", repo_full);
            } else {
                println!("Skipped repo deletion");
            }
        }
    }

    println!("Done. Mailbox '{}' removed.", name);
    Ok(())
}

/// Remove mailbox and routing entries from .corky.toml.
fn remove_from_config(name: &str) -> Result<()> {
    let config_path = resolve::corky_toml();
    if !config_path.exists() {
        return Ok(());
    }

    let content = std::fs::read_to_string(&config_path)?;
    let mut doc = content.parse::<toml_edit::DocumentMut>()?;

    // Remove [mailboxes.{name}]
    if let Some(mailboxes) = doc.get_mut("mailboxes") {
        if let Some(table) = mailboxes.as_table_mut() {
            table.remove(name);
        }
    }

    // Remove routing entries that point to this mailbox
    let mb_path = format!("mailboxes/{}", name);
    if let Some(routing) = doc.get_mut("routing") {
        if let Some(table) = routing.as_table_mut() {
            let keys_to_check: Vec<String> = table.iter().map(|(k, _)| k.to_string()).collect();
            for key in keys_to_check {
                if let Some(item) = table.get_mut(&key) {
                    if let Some(arr) = item.as_array_mut() {
                        arr.retain(|v| v.as_str() != Some(&mb_path));
                        if arr.is_empty() {
                            table.remove(&key);
                        }
                    }
                }
            }
        }
    }

    std::fs::write(&config_path, doc.to_string())?;
    println!("Removed '{}' from .corky.toml", name);

    Ok(())
}
